File: web/app/types/marker.ts
export interface Marker {
    id: number;
    user_id: number;
    scene_id: number;
    timestamp: number;
    label: string;
    color: string;
    thumbnail_path: string;
    created_at: string;
    updated_at: string;
    tags?: MarkerTagInfo[];
}

export interface CreateMarkerRequest {
    timestamp: number;
    label?: string;
    color?: string;
}

export interface UpdateMarkerRequest {
    timestamp?: number;
    label?: string;
    color?: string;
}

export interface MarkerLabelSuggestion {
    label: string;
    count: number;
}

export interface MarkerLabelGroup {
    label: string;
    count: number;
    thumbnail_marker_id: number;
}

export interface MarkerWithScene extends Marker {
    scene_title: string;
    tags?: MarkerTagInfo[];
}

// Tag info with metadata about source
export interface MarkerTagInfo {
    id: number;
    name: string;
    color: string;
    is_from_label: boolean;
}

// API Response Types
export interface MarkersResponse {
    markers: Marker[];
}

export interface LabelSuggestionsResponse {
    labels: MarkerLabelSuggestion[];
}

export interface PaginatedResponse<T> {
    data: T[];
    pagination: {
        page: number;
        limit: number;
        total_items: number;
        total_pages: number;
    };
}

// Request types for tag operations
export interface SetLabelTagsRequest {
    tag_ids: number[];
}

export interface SetMarkerTagsRequest {
    tag_ids: number[];
}

// Response types for tag operations
export interface LabelTagsResponse {
    tags: import('~/types/tag').Tag[];
}

export interface MarkerTagsResponse {
    tags: MarkerTagInfo[];
}

File: internal/api/v1/request/marker.go
package request

type CreateMarkerRequest struct {
	Timestamp int    `json:"timestamp" binding:"min=0"`
	Label     string `json:"label"`
	Color     string `json:"color"`
}

type UpdateMarkerRequest struct {
	Timestamp *int    `json:"timestamp,omitempty"`
	Label     *string `json:"label,omitempty"`
	Color     *string `json:"color,omitempty"`
}

type SetLabelTagsRequest struct {
	TagIDs []uint `json:"tag_ids" binding:"required"`
}

type SetMarkerTagsRequest struct {
	TagIDs []uint `json:"tag_ids" binding:"required"`
}
